src/DiagonalHessianApproximation.jl

```diff
@@ -0,0 +1,217 @@
+Implementation a diagonal QN method coming from the following article : 
+Andrei, N. 
+A diagonal quasi-Newton updating method for unconstrained optimization. 
+Numer Algor 81, 575–590 (2019). 
+https://doi.org/10.1007/s11075-018-0562-7
+"""
+# core structure
+mutable struct DiagonalQN{T <: Real, I <: Integer, V <: AbstractVector{T}} <: AbstractDiagonalQuasiNewtonOperator{T} 
+  d::V # Diagonal of the operator matrix
+  nrow::I
+  ncol::I
+  symmetric::Bool
+  hermitian::Bool
+  prod!
+  tprod!
+  ctprod!
+  nprod::I
+  ntprod::I
+  nctprod::I
+  args5::Bool
+  use_prod5!::Bool # true for 5-args mul! and for composite operators created with operators that use the 3-args mul!
+  Mv5::V
+  Mtu5::V
+  allocated5::Bool # true for 5-args mul!, false for 3-args mul! until the vectors are allocated
+end
+
+# constructor
+DiagonalQN(d::AbstractVector{T}) where {T <: Real} = 
+  DiagonalQN(
+    d,
+    length(d),
+    length(d),
+    true, 
+    true,  
+    (res, v, α, β) -> mulSquareOpDiagonal!(res, d, v, α, β), 
+    (res, v, α, β) -> mulSquareOpDiagonal!(res, d, v, α, β), 
+    (res, v, α, β) -> mulSquareOpDiagonal!(res, d, v, α, β), 
+    0,
+    0,
+    0,
+    true,
+    true,
+    typeof(d)(undef,0),
+    typeof(d)(undef,0),
+    true)
+
+# update function
+# s = x_{k+1} - x_k
+# y = ∇f(x_{k+1}) - ∇f(x_k)
+function push!(
+  B::DiagonalQN{T,I,V},
+  s::V,
+  y::V
+  ) where {T <: Real, I <: Integer, V <: AbstractVector{T}}
+  trA2 = 0
+  for i in eachindex(s)
+    trA2 += s[i]^4
+  end
+  sT_s = dot(s,s)
+  sT_y = dot(s,y)
+  Bs = B*s
+  sT_B_s = dot(s,Bs)
+  q = (sT_y + sT_s - sT_B_s)/trA2
+  B.d .+= q .* s.^2 .- 1
+end
+
+"""
+Implementation a spectral gradient method coming from the following article : 
+Birgin, E. G., Martínez, J. M., & Raydan, M. (2014). 
+Spectral Projected Gradient Methods: Review and Perspectives. 
+Journal of Statistical Software, 60(3), 1–21. 
+https://doi.org/10.18637/jss.v060.i03
+"""
+
+# core structure
+mutable struct SpectralGradient{T <: Real, I <: Integer, V <: AbstractVector{T}} <: AbstractDiagonalQuasiNewtonOperator{T} 
+  d::T # Diagonal coefficient of the operator matrix (multiple of identity)
+  nrow::I
+  ncol::I
+  symmetric::Bool
+  hermitian::Bool
+  prod!
+  tprod!
+  ctprod!
+  nprod::I
+  ntprod::I
+  nctprod::I
+  args5::Bool
+  use_prod5!::Bool # true for 5-args mul! and for composite operators created with operators that use the 3-args mul!
+  Mv5::V
+  Mtu5::V
+  allocated5::Bool # true for 5-args mul!, false for 3-args mul! until the vectors are allocated
+end
+
+# constructor
+SpectralGradient(d::T, n::I) where {T <: Real, I <: Integer} = 
+  SpectralGradient(
+    d,
+    n,
+    n,
+    true, 
+    true,  
+    (res, v, α, β) -> mulSquareOpDiagonal!(res, d*ones(n), v, α, β), 
+    (res, v, α, β) -> mulSquareOpDiagonal!(res, d*ones(n), v, α, β), 
+    (res, v, α, β) -> mulSquareOpDiagonal!(res, d*ones(n), v, α, β), 
+    0,
+    0,
+    0,
+    true,
+    true,
+    typeof(d*ones(n))(undef,0),
```

Comment:
This is a bit annoying. What if the user wanted to work with GPU vectors? By default `ones()` creates CPU vectors. It may be better to take `V` as input to the constructor in this case.